File: pagerank/pagerank.go
package main

import (
	"bufio"
	"bytes"
	"compress/gzip"
	"flag"
	"fmt"
	"io"
	"log"
	"math"
	"os"
	"strconv"
	"strings"
)

type Rows struct {
	Rows []*Row
}

type Row struct {
	Vals []float64
	Dims []int
}

func (r *Rows) reserve(id int) {
	for id >= len(r.Rows) {
		r.Rows = append(r.Rows, &Row{})
	}
}

func (r *Rows) Add(frid int, toids []int) {
	n := len(toids)
	v := 1.0 / float64(n)
	r.reserve(frid)
	for i := 0; i < n; i++ {
		r.reserve(toids[i])
		row := r.Rows[toids[i]]
		row.Vals = append(row.Vals, v)
		row.Dims = append(row.Dims, frid)
	}
}

func DimFor(nameToDim map[int]int, dims *[]int, name int) int {
	eid, has := nameToDim[name]
	if has {
		return eid
	}

	eid = len(*dims)
	nameToDim[name] = eid
	*dims = append(*dims, name)
	return eid
}

func ReadMatrix(filename string, rows *Rows) ([]int, error) {
	r, err := os.Open(filename)
	if err != nil {
		return nil, err
	}
	defer r.Close()

	gr, err := gzip.NewReader(r)
	if err != nil {
		return nil, err
	}
	defer gr.Close()

	br := bufio.NewReader(gr)

	var buf bytes.Buffer
	var dims []int
	nameToDim := map[int]int{}

	frId := -1
	var toIds []int

	for {
		b, p, err := br.ReadLine()
		if err == io.EOF {
			if len(toIds) > 0 {
				rows.Add(frId, toIds)
			}
			return dims, nil
		} else if err != nil {
			return nil, err
		}

		buf.Write(b)
		if p {
			continue
		}

		line := buf.String()
		buf.Reset()

		if strings.HasPrefix(line, "#") {
			continue
		}

		ix := strings.Index(line, "\t")
		if ix < 0 {
			return nil, fmt.Errorf("invalid line: %s", line)
		}

		ln, err := strconv.ParseInt(line[:ix], 10, 64)
		if err != nil {
			return nil, err
		}

		rn, err := strconv.ParseInt(line[ix+1:], 10, 64)
		if err != nil {
			return nil, err
		}

		lid := DimFor(nameToDim, &dims, int(ln))
		rid := DimFor(nameToDim, &dims, int(rn))

		if lid == frId || frId < 0 {
			frId = lid
			toIds = append(toIds, rid)
			continue
		}

		rows.Add(frId, toIds)
		frId = lid
		toIds = toIds[:0]
		toIds = append(toIds, rid)
	}
}

func ComputeL1Norm(a, b []float64) float64 {
	s := 0.0
	for i, n := 0, len(a); i < n; i++ {
		s += math.Abs(a[i] - b[i])
	}
	return s
}

func Rank(rows *Rows, beta, epsilon float64) []float64 {
	n := len(rows.Rows)
	a := make([]float64, n)
	b := make([]float64, n)

	v := 1.0 / float64(n)
	for i := 0; i < n; i++ {
		a[i] = v
	}

	for {

		for i := 0; i < n; i++ {
			b[i] = 0.0
		}

		s := 0.0

		for i := 0; i < n; i++ {
			row := rows.Rows[i]
			for j, m := 0, len(row.Dims); j < m; j++ {
				b[i] += a[row.Dims[j]] * row.Vals[j] * beta
			}

			s += b[i]
		}

		z := (1.0 - s) / float64(n)
		for i := 0; i < n; i++ {
			b[i] += z
		}

		d := ComputeL1Norm(a, b)
		if d < epsilon {
			return b
		}

		a, b = b, a
	}
}

func main() {
	flag.Parse()

	if flag.NArg() != 1 {
		fmt.Fprintln(os.Stderr, "usage: pr file.gz")
		os.Exit(1)
	}

	var rows Rows
	dims, err := ReadMatrix(flag.Arg(0), &rows)
	if err != nil {
		log.Panic(err)
	}

	r := Rank(&rows, 0.8, 1e-10)

	for i, n := 0, len(r); i < n; i++ {
		fmt.Printf("% 6d: %0.3e\n", dims[i], r[i])
	}
}

File: pagerank/pagerank.cc

#include <memory>
#include <stdio.h>
#include <unistd.h>
#include <sys/fcntl.h>
#include <string>
#include <unordered_map>
#include <math.h>
#include <zlib.h>

#include "status.h"
#include "util.h"

namespace {

typedef std::unordered_map<int, int> IdMap;

class Row {
 public:
  std::vector<double> vals;
  std::vector<int> dims;
 private:
};

class Rows {
 public:
  Rows() {
    rows_.reserve(500000);
  }

  ~Rows() {
    for (int i = 0, n = rows_.size(); i < n; i++) {
      delete rows_[i];
    }
  }

  Row* get(int i) {
    return rows_[i];
  }

  size_t Size() {
    return rows_.size();
  }

  void Add(int frid, std::vector<int>& toid) {
    double f = 1.0 / (double)toid.size();
    ReserveFor(frid);
    for (int i = 0, n = toid.size(); i < n; i++) {
      ReserveFor(toid[i]);
      Row* r = rows_[toid[i]];
      r->vals.push_back(f);
      r->dims.push_back(frid);
    }
  }
 private:
  void ReserveFor(int id) {
    if (id < rows_.size()) {
      return;
    }
    // rows_.reserve(id + 1);
    while (id >= rows_.size()) {
      rows_.push_back(new Row);
    }
  }

  std::vector<Row*> rows_;
};


int DimFor(IdMap* name_to_dim, std::vector<int>* dims, int name) {

  std::pair<IdMap::iterator, bool> r = name_to_dim->insert(
    IdMap::value_type(name, dims->size()));
  if (r.second) {
    dims->push_back(name);
  }

  return r.first->second;
}

Status ReadGraph(const char* filename, Rows* rows, std::vector<int>* dims) {

  IdMap name_to_dim;

  gzFile fd = gzopen(filename, "rb");
  if (fd == 0) {
    return ERR("unable to open file");
  }

  std::string line;

  int frId = -1;
  std::vector<int> toIds;

  static const size_t BUFFER_SIZE = 16*1024;
  char buf[BUFFER_SIZE + 1];
  while (size_t n = gzread(fd, buf, BUFFER_SIZE)) {
    if (n == (size_t)-1) {
      return ERR("read error");
    }

    if (!n) {
      break;
    }

    char* p = buf;
    for (;;) {
      char* c = (char*)memchr(p, '\n', (buf + n) - p);
      if (c == NULL) {
        line.assign(p, (buf + n) - p);
        break;
      }

      line.append(p, (size_t)(c - p));
      p = c + 1;

      if (line[0] == '#') {
        line.clear();
        continue;
      }

      size_t ix = line.find('\t');
      if (ix == std::string::npos) {
        return ERR("invalid line");
      }

      int ln, rn;

      std::string lns(line.substr(0, ix));
      Status did = util::ParseInt(lns, 10, &ln);
      if (!did.ok()) {
        return did;
      }

      // take the stupid \r off the end.
      int e = line.size() - 1;
      while (line[e] == '\r') {
        e--;
      }

      std::string rns(line.substr(ix + 1, e-ix));
      did = util::ParseInt(rns, 10, &rn);
      if (!did.ok()) {
        return did;
      }

      line.clear();
      ln = DimFor(&name_to_dim, dims, ln);
      rn = DimFor(&name_to_dim, dims, rn);

      if (ln == frId || frId < 0) {
        frId = ln;
        toIds.push_back(rn);
        continue;
      }

      rows->Add(frId, toIds);
      frId = ln;
      toIds.clear();
      toIds.push_back(rn);
    }
  }

  if (!toIds.empty()) {
    rows->Add(frId, toIds);
  }

  gzclose(fd);
  return NoErr();
}

double ComputeL1Norm(double* a, double* b, size_t n) {
  double s = 0.0;
  for (int i = 0; i < n; i++) {
    s += fabs(a[i] - b[i]);
  }
  return s;
}

double* Rank(Rows& rows, double beta, double epsilon) {
  size_t n = rows.Size();
  double* a = new double[n];
  double* b = new double[n];

  double v = 1.0 / (double)n;
  for (int i = 0; i < n; i++) {
    a[i] = v;
  }

  for (;;) {
    memset(b, 0, sizeof(double)*n);

    double s = 0.0;

    for (int i = 0; i < n; i++) {
      Row* row = rows.get(i);
      for (int j = 0, m = row->dims.size(); j < m; j++) {
        b[i] += a[row->dims[j]] * row->vals[j] * beta;
      }
      s += b[i];
    }

    s = (1.0 - s) / (double)n;
    for (int i = 0; i < n; i++) {
      b[i] += s;
    }

    double d = ComputeL1Norm(a, b, n);
    if (d < epsilon) {
      delete[] a;
      return b;
    }

    double* t = a;
    a = b;
    b = t;
  }
}

} // anonymous

int main(int argc, char* argv[]) {
  Rows rows;
  std::vector<int> dims;

  if (argc < 2) {
    fprintf(stderr, "usage: %s file.gz\n", argv[0]);
    return 1;
  }

  Status did = ReadGraph(argv[1], &rows, &dims);
  if (!did.ok()) {
    fprintf(stderr, "ERROR %s\n", did.what());
    return 1;
  }

  double* r = Rank(rows, 0.8, 1e-10);

  for (int i = 0, n = dims.size(); i < n; i++) {
    printf("%*d: %0.3e\n", 6, dims[i], r[i]);
  }

  return 0;
}